System/Win32/WindowsString/Types.hsc

```diff
@@ -88,11 +104,38 @@ newTString     :: WindowsString -> IO LPCTSTR
 -- UTF-16 version:
 -- the casts are from 'Ptr Word16' to 'Ptr CWchar', which is safe
 withTString (WindowsString str) f    = useAsCWString str (\ptr -> f (castPtr ptr))
+withFilePath path = useAsCWStringSafe path
 withTStringLen (WindowsString str) f = useAsCWStringLen str (\(ptr, len) -> f (castPtr ptr, len))
 peekTString    = fmap WindowsString . packCWString . castPtr
 peekTStringLen = fmap WindowsString . packCWStringLen . first castPtr
 newTString (WindowsString str) = fmap castPtr $ newCWString str
 
+foreign import ccall unsafe "wchar.h wcslen" c_wcslen
+    :: CWString -> IO SIZE_T
+
+-- | Wrapper around 'useAsCString', checking the encoded 'FilePath' for internal NUL octets as these are
+-- disallowed in Windows filepaths. See https://gitlab.haskell.org/ghc/ghc/-/issues/13660
+useAsCWStringSafe :: WindowsPath -> (CWString -> IO a) -> IO a
+useAsCWStringSafe wp@(WS path) f = useAsCWString path $ \(castPtr -> ptr) -> do
+    let len = SBS.numWord16 path
+    clen <- c_wcslen ptr
+    if clen == fromIntegral len
+        then f ptr
+        else do
+          path' <- either (const (_toStr wp)) id <$> (EX.try @IOException) (decodeWithBaseWindows path)
+          ioError (err path')

```

Comment:
This is a little more efficient than the String variant and we do the same in unix package.